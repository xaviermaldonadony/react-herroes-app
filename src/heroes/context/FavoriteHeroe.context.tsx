import type { Hero } from '../types/hero.interface';
import {
  createContext,
  useEffect,
  useState,
  type PropsWithChildren,
} from 'react';

interface FavoriteHeroContext {
  // state
  favorites: Hero[];
  favoriteCount: number;

  //   methods
  isFavorite: (hero: Hero) => boolean;
  toggleFavorites: (hero: Hero) => void;
}
// eslint-disable-next-line react-refresh/only-export-components
export const FavoriteHeroContext = createContext({} as FavoriteHeroContext);

const getFavoritesFromLocalStorage = (): Hero[] => {
  const favorites = localStorage.getItem('favorites');
  return favorites ? JSON.parse(favorites) : [];
};

export const FavoriteHeroeProvider = ({ children }: PropsWithChildren) => {
  const [favorites, setFavorites] = useState<Hero[]>(
    getFavoritesFromLocalStorage()
  );

  const toggleFavorites = (hero: Hero) => {
    const heroExist = favorites.find((h) => h.id === hero.id);

    if (heroExist) {
      setFavorites(favorites.filter((h) => h.id !== hero.id));
      return;
    }

    setFavorites([...favorites, hero]);
  };

  const isFavorite = (hero: Hero) => favorites.some((h) => h.id === hero.id);

  useEffect(() => {
    localStorage.setItem('favorites', JSON.stringify(favorites));
  }, [favorites]);

  return (
    <FavoriteHeroContext
      value={{
        favoriteCount: favorites.length,
        favorites,

        // methods
        isFavorite,
        toggleFavorites,
      }}
    >
      {children}
    </FavoriteHeroContext>
  );
};
