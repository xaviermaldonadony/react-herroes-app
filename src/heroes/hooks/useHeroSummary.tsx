import { useQuery } from '@tanstack/react-query';
import { getSummaryAcion } from '../actions/get-summary-actions';

export const useHeroSummary = () => {
  return useQuery({
    queryKey: ['summary-information'],
    queryFn: getSummaryAcion,
    staleTime: 1000 * 60 * 5, // 5 mins
  });
};
