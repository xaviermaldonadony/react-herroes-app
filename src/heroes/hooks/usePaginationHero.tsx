import { useQuery } from '@tanstack/react-query';
import { getHeroesByPageAction } from '../actions/get-heroes-by-page.action';

export const usePaginationHero = (
  page: number,
  limit: number,
  category = 'all'
) => {
  return useQuery({
    queryKey: ['heroes', { page, limit, category }],
    queryFn: () => getHeroesByPageAction(+page, +limit, category),
    staleTime: 1000 * 60 * 5, // 5 minutes
  });
};
