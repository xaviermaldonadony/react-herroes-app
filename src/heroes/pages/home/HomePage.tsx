import { useState } from 'react';

import { Tabs, TabsList, TabsTrigger, TabsContent } from '@/components/ui/tabs';
import { CustomJumbotron } from '@/components/custom/CustomJumbotron';
import { HeroStats } from '@/heroes/components/HeroStats';
import { HeroGrid } from '@/heroes/components/HeroGrid';
import { CustomPagination } from '@/components/custom/CustomPagination';
import { CustomBreadcrumbs } from '@/components/custom/CustomBreadcrumbs';

export const HomePage = () => {
  const [activetab, setActivetab] = useState<
    'heroes' | 'villains' | 'favorites' | 'all'
  >('all');
  return (
    <>
      <>
        <CustomBreadcrumbs currentPage='Super Heroes' breadcrumbs={[]} />
        {/* Header */}
        <CustomJumbotron
          title={'Super Heroe Universe'}
          description='
          Discover, explore, and manage your favorite superheroes and villains'
        />

        {/* Stats Dashboard */}
        <HeroStats />

        {/* Tabs */}
        {/* To do  */}
        <Tabs value={activetab} className='mb-8'>
          <TabsList className='grid w-full grid-cols-4'>
            <TabsTrigger value='all' onClick={() => setActivetab('all')}>
              All Characters (16)
            </TabsTrigger>
            <TabsTrigger
              value='favorites'
              className='flex items-center gap-2'
              onClick={() => setActivetab('favorites')}
            >
              Favorites (3)
            </TabsTrigger>
            <TabsTrigger value='heroes' onClick={() => setActivetab('heroes')}>
              Heroes (12)
            </TabsTrigger>
            <TabsTrigger
              value='villains'
              onClick={() => setActivetab('villains')}
            >
              Villains (2)
            </TabsTrigger>
          </TabsList>
          <TabsContent value='all'>
            {/* Character Grid */}
            {/* Show all characters */}
            <HeroGrid />
          </TabsContent>
          <TabsContent value='favorites'>
            {/* TODO Show all favorites */}
            <h1>Favorites</h1>
            <HeroGrid />
          </TabsContent>
          <TabsContent value='heroes'>
            {/* TODO Show all Heroes*/}
            <h1>Heroes</h1>
            <HeroGrid />
          </TabsContent>
          <TabsContent value='villains'>
            {/* TODO Show all Villains*/}
            <h1>Villains</h1>
            <HeroGrid />
          </TabsContent>
        </Tabs>

        {/* Pagination */}
        <CustomPagination totalPages={8} />
      </>
    </>
  );
};
