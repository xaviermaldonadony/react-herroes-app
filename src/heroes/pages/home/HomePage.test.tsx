import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, test, vi } from 'vitest';
import { HomePage } from './HomePage';
import { MemoryRouter } from 'react-router';
import { usePaginationHero } from '@/heroes/hooks/usePaginationHero';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { beforeEach } from 'node:test';
import { FavoriteHeroeProvider } from '@/heroes/context/FavoriteHeroe.context';

vi.mock('@/heroes/hooks/usePaginationHero');

const mockUsePaginationdHero = vi.mocked(usePaginationHero);

mockUsePaginationdHero.mockReturnValue({
  data: [],
  isLoading: false,
  isError: false,
  isSuccess: true,
} as unknown as ReturnType<typeof usePaginationHero>);

const queryClient = new QueryClient();

const renderHomePage = (initialEntries: string[] = ['/']) => {
  return render(
    <MemoryRouter initialEntries={initialEntries}>
      <FavoriteHeroeProvider>
        <QueryClientProvider client={queryClient}>
          <HomePage />
        </QueryClientProvider>
      </FavoriteHeroeProvider>
    </MemoryRouter>
  );
};

describe('HomePage', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('Should render HomePage with default values', () => {
    const { container } = renderHomePage();
    expect(container).toMatchSnapshot();
  });

  test('Should call usePaginationHero with default values', () => {
    renderHomePage();
    expect(mockUsePaginationdHero).toHaveBeenCalledWith(1, 6, 'all');
  });

  test('Should call usePaginationHero with custom query params', () => {
    renderHomePage(['/?page=2&limit=10&category=villains']);
    const lastCall = mockUsePaginationdHero.mock.lastCall;

    // Assert that the last call's arguments match the custom query params
    expect(lastCall).toEqual([2, 10, 'villains']);
  });

  test('Should call usePaginationHero with default page and same limit on tab clicked', () => {
    renderHomePage(['/?tab=favorites&page=2&limit=10']);
    const [, , , villainsTab] = screen.getAllByRole('tab');

    fireEvent.click(villainsTab);
    const lastCall = mockUsePaginationdHero.mock.lastCall;
    expect(lastCall).toEqual([1, 10, 'villain']);
  });
});
