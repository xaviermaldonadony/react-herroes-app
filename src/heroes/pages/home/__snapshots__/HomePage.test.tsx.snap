// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomePage > Should render HomePage with default values 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors text-indigo-800 font-semibold"
          data-slot="breadcrumb-link"
        >
          Super Heroes
        </a>
      </li>
    </ol>
  </nav>
  <div
    class="text-center mb-8"
  >
    <h1
      class="text-5xl font-bold bg-gradient-to-r from-blue-600 to-purple-600 bg-clip-text text-transparent mb-4"
    >
      Super Heroes Universe
    </h1>
    <p
      class="text-gray-600 text-lg"
    >
       Discover, explore, and manage your favorite superheroes and villains
    </p>
  </div>
  <div>
    Loading...
  </div>
  <div
    class="flex flex-col gap-2 mb-8"
    data-orientation="horizontal"
    data-slot="tabs"
    dir="ltr"
  >
    <div
      aria-orientation="horizontal"
      class="bg-muted text-muted-foreground h-9 items-center justify-center rounded-lg p-[3px] grid w-full grid-cols-4"
      data-orientation="horizontal"
      data-slot="tabs-list"
      role="tablist"
      style="outline: none;"
      tabindex="0"
    >
      <button
        aria-controls="radix-«r0»-content-all"
        aria-selected="true"
        class="data-[state=active]:bg-background dark:data-[state=active]:text-foreground focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:outline-ring dark:data-[state=active]:border-input dark:data-[state=active]:bg-input/30 text-foreground dark:text-muted-foreground inline-flex h-[calc(100%-1px)] flex-1 items-center justify-center gap-1.5 rounded-md border border-transparent px-2 py-1 text-sm font-medium whitespace-nowrap transition-[color,box-shadow] focus-visible:ring-[3px] focus-visible:outline-1 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:shadow-sm [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-slot="tabs-trigger"
        data-state="active"
        id="radix-«r0»-trigger-all"
        role="tab"
        tabindex="-1"
        type="button"
      >
        All Characters 
      </button>
      <button
        aria-controls="radix-«r0»-content-favorites"
        aria-selected="false"
        class="data-[state=active]:bg-background dark:data-[state=active]:text-foreground focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:outline-ring dark:data-[state=active]:border-input dark:data-[state=active]:bg-input/30 text-foreground dark:text-muted-foreground h-[calc(100%-1px)] flex-1 justify-center rounded-md border border-transparent px-2 py-1 text-sm font-medium whitespace-nowrap transition-[color,box-shadow] focus-visible:ring-[3px] focus-visible:outline-1 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:shadow-sm [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 flex items-center gap-2"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-slot="tabs-trigger"
        data-state="inactive"
        id="radix-«r0»-trigger-favorites"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Favorites (
        0
        )
      </button>
      <button
        aria-controls="radix-«r0»-content-heroes"
        aria-selected="false"
        class="data-[state=active]:bg-background dark:data-[state=active]:text-foreground focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:outline-ring dark:data-[state=active]:border-input dark:data-[state=active]:bg-input/30 text-foreground dark:text-muted-foreground inline-flex h-[calc(100%-1px)] flex-1 items-center justify-center gap-1.5 rounded-md border border-transparent px-2 py-1 text-sm font-medium whitespace-nowrap transition-[color,box-shadow] focus-visible:ring-[3px] focus-visible:outline-1 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:shadow-sm [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-slot="tabs-trigger"
        data-state="inactive"
        id="radix-«r0»-trigger-heroes"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Heroes 
      </button>
      <button
        aria-controls="radix-«r0»-content-villains"
        aria-selected="false"
        class="data-[state=active]:bg-background dark:data-[state=active]:text-foreground focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:outline-ring dark:data-[state=active]:border-input dark:data-[state=active]:bg-input/30 text-foreground dark:text-muted-foreground inline-flex h-[calc(100%-1px)] flex-1 items-center justify-center gap-1.5 rounded-md border border-transparent px-2 py-1 text-sm font-medium whitespace-nowrap transition-[color,box-shadow] focus-visible:ring-[3px] focus-visible:outline-1 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:shadow-sm [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-slot="tabs-trigger"
        data-state="inactive"
        id="radix-«r0»-trigger-villains"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Villains 
      </button>
    </div>
    <div
      aria-labelledby="radix-«r0»-trigger-all"
      class="flex-1 outline-none"
      data-orientation="horizontal"
      data-slot="tabs-content"
      data-state="active"
      id="radix-«r0»-content-all"
      role="tabpanel"
      style="animation-duration: 0s;"
      tabindex="0"
    >
      <div
        class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-8 mb-8"
      />
    </div>
    <div
      aria-labelledby="radix-«r0»-trigger-favorites"
      class="flex-1 outline-none"
      data-orientation="horizontal"
      data-slot="tabs-content"
      data-state="inactive"
      hidden=""
      id="radix-«r0»-content-favorites"
      role="tabpanel"
      tabindex="0"
    />
    <div
      aria-labelledby="radix-«r0»-trigger-heroes"
      class="flex-1 outline-none"
      data-orientation="horizontal"
      data-slot="tabs-content"
      data-state="inactive"
      hidden=""
      id="radix-«r0»-content-heroes"
      role="tabpanel"
      tabindex="0"
    />
    <div
      aria-labelledby="radix-«r0»-trigger-villains"
      class="flex-1 outline-none"
      data-orientation="horizontal"
      data-slot="tabs-content"
      data-state="inactive"
      hidden=""
      id="radix-«r0»-content-villains"
      role="tabpanel"
      tabindex="0"
    />
  </div>
  <div
    class="flex items-center justify-center space-x-2"
  >
    <button
      class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5"
      data-slot="button"
      disabled=""
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-chevron-left h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m15 18-6-6 6-6"
        />
      </svg>
      Previous
    </button>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5"
      data-slot="button"
    >
      Next
      <svg
        aria-hidden="true"
        class="lucide lucide-chevron-right h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m9 18 6-6-6-6"
        />
      </svg>
    </button>
  </div>
</div>
`;
